File: rust/src/maze_solver.rs
pub type Point = (i32, i32);

static DIR: [(i32, i32); 4] = [(-1, 0), (0, -1), (1, 0), (0, 1)];

fn walk(
    maze: &Vec<Vec<char>>,
    wall: char,
    curr: Point,
    end: Point,
    seen: &mut Vec<Vec<bool>>,
    path: &mut Vec<Point>,
) -> bool {
    let (x, y) = curr;

    if x < 0 || x >= maze[0].len() as i32 || y < 0 || y >= maze.len() as i32 {
        return false;
    }

    if maze[y as usize][x as usize] == wall {
        return false;
    }

    if curr == end {
        path.push(end);
        return true;
    }

    if seen[y as usize][x as usize] {
        return false;
    }

    seen[y as usize][x as usize] = true;
    path.push(curr);

    for (dx, dy) in DIR.iter() {
        let next_point = (x + dx, y + dy);
        if walk(maze, wall, next_point, end, seen, path) {
            return true;
        }
    }

    path.pop();

    false
}

pub fn solve(maze: Vec<Vec<char>>, wall: char, start: Point, end: Point) -> Vec<Point> {
    let mut seen: Vec<Vec<bool>> = Vec::new();
    let mut path: Vec<Point> = Vec::new();

    for _ in 0..maze.len() {
        seen.push(vec![false; maze[0].len()]);
    }

    walk(&maze, wall, start, end, &mut seen, &mut path);
    path
}

File: java/src/MazeSolver.java
package src;

import java.util.ArrayList;
import java.util.List;
// import src.Point;

class MazePoint {
    int x, y;

    public MazePoint(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class MazeSolver {
    private static final int[][] dir = {
        {-1, 0}, // left
        {0, -1}, // up
        {1, 0},  // right
        {0, 1},  // down
    };

    private static boolean walk(String[] maze, String wall, MazePoint curr, MazePoint end, boolean[][] seen, List<MazePoint> path) {
        if (curr.x < 0 || curr.x >= maze[0].length() || curr.y < 0 || curr.y >= maze.length) {
            return false;
        }

        if (maze[curr.y].charAt(curr.x) == wall.charAt(0)) {
            return false;
        }

        if (curr.x == end.x && curr.y == end.y) {
            path.add(end);
            return true;
        }

        if (seen[curr.y][curr.x]) {
            return false;
        }

        seen[curr.y][curr.x] = true;
        path.add(curr);

        for (int i = 0; i < dir.length; ++i) {
            int x = dir[i][0];
            int y = dir[i][1];
            if (walk(maze, wall, new MazePoint(curr.x + x, curr.y + y), end, seen, path)) {
                return true;
            }
        }

        path.remove(path.size() - 1);

        return false;
    }

    public static List<MazePoint> solve(String[] maze, String wall, MazePoint start, MazePoint end) {
        boolean[][] seen = new boolean[maze.length][maze[0].length()];
        List<MazePoint> path = new ArrayList<>();

        walk(maze, wall, start, end, seen, path);
        return path;
    }
}
